Add a table widget displaying the status of ten agencies

issyl0 opened the pull request:
![screen shot 2014-03-20 at 13 27 54](https://f.cloud.github.com/assets/355033/2472370/865d0250-b034-11e3-8217-424626502d9b.png)

Commit:
Add the list of top ten agencies and their statuses

- The Ruby code that retrieves the data works but needs *desperately*
  refactoring, which I'll tackle.
- All of the data isn't here yet - transition managers also want
  their RAG status and the transition target date - but this proves
  that it's possible.
- Resize the existing widgets to make room for this one on the same
  line.
- Reduce the font-size on the list widget a bit so that the agency
  names don't overlap with their statuses.

Commit:
By CamelCase the status, I meant Titleize. Actually do it now.

Commit:
Refactor the data retreival for the top ten transitioning orgs table

Commit:
Fix an off-by-one error

- This removes the spreadsheet column headers from the list widget.
- Fix terminology in a comment.

Commit:
Add the target transition month and RAG status to the widget

Commit:
Name the top_ten_agencies hash keys according to the value description

- Not just non-descript "value".

Commit:
Fix the org_top_ten list alignment with an HTML table and some CSS

Commit:
Ignore the potentially present "- live" in the target month data

- There's currently a "February - live" target month for UKBA - it
  looks messy in the table.

issyl0 commented:
At the request of transition managers, this adds a widget to the dashboard showing the organisation/agency name, red/amber/green status, target transition month, and the phase of processing they're in.

issyl0 commented:
Rebased against master and fixed the conflicts, so someone can merge this now. Thanks @jennyd for the rebase tutorial!

Commit:
Remove the abbreviation for "business as usual" for ease of reading

- Discussed with William Reddin.

issyl0 commented:
This dashboard widget actually now looks like: 
![screen shot 2014-03-20 at 13 25 14](https://f.cloud.github.com/assets/355033/2472377/927a6730-b034-11e3-990b-8c6aeff2ebf7.png)

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,37 @@
+end
+
+SCHEDULER.every '2h', :first_in => 0 do |job|
+  row = 2
```
This looks like a magic number. It also has a fixed value. I would make it a constant, give it a descriptive name and move it next to the spreadsheet URL as it is about the position of something in that spreadsheet.

issyl0 replied:
The `row` variable is spreadsheet-specific, but it is incremented for each of the 10 agencies to retrieve the correct data from the correct rows (from row two onwards to avoid getting spreadsheet headers). Making its value a constant wouldn't let it be incremented?

jamiecobbett replied:
Woops, my mistake :)

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,37 @@
+  org_number = 0
+  (0..9).each do |orgs|
+    org = XmlSimple.xml_in(RestClient.get("#{spreadsheet_end_url}/R#{row}C1"))['content']['content']
+    status = abbreviate_status(XmlSimple.xml_in(RestClient.get("#{spreadsheet_end_url}/R#{row}C2"))['content']['content'])
```
These lines are really long. We should keep to 80 lines if we can, as per the styleguide: https://github.com/alphagov/styleguides/blob/master/ruby.md

There's lots of repetition here too. What about a method:

```
def get_cell_content_xml(row, column)
  XmlSimple.xml_in(RestClient.get("#{spreadsheet_end_url}/R#{row}#{column}"))
end
```

issyl0 replied:
Yeah, repetition, but be glad you didn't have to look at the original version in this review! ;-)

issyl0 replied:
Refactored. There are still two long lines, but I can't really get around those. I thought it best to condense the spreadsheet URL into one string to avoid confusion - it's still easily editable if it needs to be.

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,37 @@
+SCHEDULER.every '2h', :first_in => 0 do |job|
+  row = 2
+  org_number = 0
+  (0..9).each do |orgs|
```
`orgs` doesn't appear to be used again. If that is the case, it doesn't need to be there.

issyl0 replied:
I didn't think you could do `.each` without a `|variable|`. Would changing it to `|i|` like a counter be acceptable, even though that counter wouldn't be used anywhere either?

jamiecobbett replied:
Yes, you can. Having any variable makes me wonder what it's for.

issyl0 replied:
Oh yes. Done.

Commit:
Remove superfluous `|orgs|` in (0..9).each block

Commit:
Further refactoring of the spreadsheet data retrieval code

Commit:
Remove unnecessary animation config in the list widget JS

- The list widget data was set to animate when it loaded, but it
  doesn't seem to be working and this - touchwood - does nothing bad
  to the widget. At least there's no potential for distracting people
  with animation now it's removed.

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,48 @@
+  # Worksheet three for the list of top 10 agencies.
+  spreadsheet_url = "https://spreadsheets.google.com/feeds/cells/#{ENV['main_spreadsheet_key']}/3/public/values"
+
+  data = Hash.new()
```
Why is this a hash if we only use the values?

issyl0 replied:
This is a hash because I followed the example of getting values from the one where the `org_top_ten` hash values are passed into the widget. You make a good point though, it should probably be an array.

issyl0 replied:
I changed `data` to be an array. Hopefully this'll be the final thing!

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,48 @@
+
+  data = Hash.new()
+  (0..4).each do |i|
+    data[i] = XmlSimple.xml_in(RestClient.get("#{spreadsheet_url}/R#{row}C#{column}"))['content']['content']
```
I think there's too much on this line. What about:

```
response = RestClient.get("#{spreadsheet_url}/R#{row}C#{column}")
data[i] = XmlSimple.xml_in(response)['content']['content']
```

issyl0 replied:
That works. Edited.

jamiecobbett commented on jobs/org_top_ten.rb:
```diff
@@ -0,0 +1,48 @@
+
+org_top_ten = Hash.new({ value: 0 })
+
+def get_spreadsheet_xml(row, column)
```
This method should be named for what it returns. 

Additionally, it returns an array of values which the order of which the caller is expected to know. What about creating a data carrier object? [Struct](http://ruby-doc.org/core-1.9.2/Struct.html) and [OpenStruct](http://ruby-doc.org/stdlib-1.9.2/libdoc/ostruct/rdoc/OpenStruct.html) can be useful for this.

The name of the method and that data carrier class might be nearly the same.

issyl0 replied:
I didn't realise that Struct or OpenStruct existed. They look useful.

They seem to work with hashes, yet https://github.com/alphagov/transition-dashboard/pull/4#discussion_r10836248 appears to suggest that `data` should not be a hash (possibly because it has numeric keys?). Numeric keys are just because I no longer assign variables to each spreadsheet field value - I can't because it's condensed into much cleaner and nicer code with no repetition. I don't understand how OpenStruct would work in this case, or how I'd go about assigning better key names to the code that exists there already.

Alternatively, I could assign a descriptive variable (`status`, `rag`, ...) to every `data[0]`, `data[1]`, ..., but that would look more messy. I think it's fine given with its keys, `data` is not used anywhere else apart from in `org_top_ten` hash creation where each field has an appropriate and descriptive label key itself, and they're assigned in order of 0 to 3.

issyl0 replied:
The data retrieval method is now named more descriptively.

Commit:
Split the data retrieval XmlSimple...RestClient... line into two

jamiecobbett commented:
I really like the way this looks :+1:

issyl0 commented:
Reference the way the widget itself looks, William Reddin suggested that I try to make the RAG status output actually coloured red, amber or green. I don't think that'd look very good with the background as it is, or even if I change the background to white. I'm happy to try it, with my limited CSS skills and no doubt some frustration, but not necessarily in this pull request!

adammaddison commented:
I :+1: 'd Will's request.  Then remembered I'm Issy's delivery manager and should not be encouraging scope creep!
So ignore that for now.  Put a ticket in the icebox.

jamiecobbett commented:
@adammaddison it's eerie how much your avatar looks like you.

Commit:
Change the `data` hash in the data retrieval method to an array

- The `data` hash only had numeric values as keys.
- Adhere to the style guide with Array.new without parentheses.
- Give the data retrieval method and a variable more descriptive
  names.

Commit:
Make the spreadsheet URL a constant

jamiecobbett committed:
Simplify spreadsheet method

jamiecobbett committed:
Refactor Spreadsheet code

adammaddison commented:
@jamiecobbett no idea what you're on about...
